src/main/java/com/bobocode/hoverla/bring/context/BeanValidator.java

```diff
@@ -1,18 +1,174 @@
+            throw new BeanValidationException(INCORRECT_BEAN_NAME.formatted(beanDefinition.type().getSimpleName()));
+        }
+    }
+
+    private void checkDependencies(BeanDefinition beanDefinition, List<BeanDefinition> beanDefinitions) {
+        log.trace("Checking dependencies for bean definition: {} - {}", beanDefinition.name(), beanDefinition.type());
+        Map<String, Class<?>> dependencies = beanDefinition.dependencies();
+        if (dependencies.isEmpty()) {
+            return;
+        }
+        LinkedHashMap<String, List<String>> requiredDependencies = new LinkedHashMap<>();
+        requiredDependencies.put(beanDefinition.name(), beanDefinition.dependencies().keySet().stream().toList());
+
+        for (var dependency : dependencies.entrySet()) {
+            log.trace("Checking dependency {} - {}", dependency.getKey(), dependency.getValue());
+            List<BeanDefinition> beansByName = findBeansByName(dependency.getKey(), beanDefinitions);
+            if (beansByName.size() > 1) {
+                throw new BeanValidationException(CONTEXT_CONTAINS_BEANS_WITH_DUPLICATE_NAMES + dependency.getKey());
+            }
+            if (beansByName.size() == 1) {
+                checkTypeMatching(dependency.getValue(), beansByName.get(0));
+                checkCircularDependency(beanDefinition, beansByName.get(0), beanDefinitions, requiredDependencies);
+            } else {
+                checkBeansByType(dependency.getValue(), beanDefinitions, dependency.getKey());
+            }
+        }
+    }
+
+    private List<BeanDefinition> findBeansByName(String source, List<BeanDefinition> allBeanDefinitions) {
+        return allBeanDefinitions.stream()
+                .filter(b -> source.equals(b.name()))
+                .toList();
+    }
+
+    private void checkTypeMatching(Class<?> requiredType, BeanDefinition foundBeanDefinition) {
+        Class<?> foundType = foundBeanDefinition.type();
+        if (!foundType.equals(requiredType)) {
+            throw new BeanValidationException(DIFFERENT_TYPES_IN_DEPENDENCIES.formatted(requiredType, foundType));
+        }
+    }
+
+    private void checkCircularDependency(BeanDefinition source,
+                                         BeanDefinition found,
+                                         List<BeanDefinition> beanDefinitions,
+                                         LinkedHashMap<String, List<String>> requiredDependencies) {
+        Map<String, Class<?>> dependencies = found.dependencies();
+        if (dependencies.isEmpty()) {
+            return;
+        }
+        if (dependencies.containsKey(source.name())) {
+            String message = buildCircularExceptionMessage(found, requiredDependencies);
+            throw new BeanValidationException(message.formatted(source.name(), found.name()));
+        }
+        for (Map.Entry<String, Class<?>> dependency : dependencies.entrySet()) {
+            if (requiredDependencies.containsKey(dependency.getKey())) {
```

Comment:
May you explain why this situation may lead to circular dependency, please? I don't get it

Reply:
requiredDependencies is a map that contains all dependencies which are needed for creating the current bean definition; 
the key is the name of the required bean;
value is a list of required dependencies for this bean.